### ISSUE TITLE
missing dependency in systemd unit file

### ISSUE THREAD

**Issue by schlitzered** [rsyslog/rsyslog-pkg-rhel-centos#72]
at least the CentOS 7 unit file has missing dependencies to the network target.

can you please add this:

```
[Unit]
Wants=network.target network-online.target
After=network.target network-online.target
```

please also check if the same is missing for CentOS 8

**Comment by ebsf1** [comment]
@schlitzered , I found the same for Ubuntu 16.04 & 18.04, and this was helpful in getting the RELP listener to bind to its socket (as discussed in https://github.com/rsyslog/rsyslog/issues/4293) and in enabling limited functionality in imuxsock rulesets (as discussed in https://github.com/rsyslog/rsyslog/issues/4299), although in each case, this surfaced other issues.

In my case, though, all that was necessary was

`After=network-online.target`

My question is, which functionality or behavior in rsyslog was enabled when you added these dependencies?

Otherwise, do you know whether this dependency prevents logging of messages before this target occurs?

**Comment by alorbach** [comment]
> at least the CentOS 7 unit file has missing dependencies to the network target.
> 
> can you please add this:
> 
> ```
> [Unit]
> Wants=network.target network-online.target
> After=network.target network-online.target
> ```

Not sure if this is good Idea for the default systemd file. 
A default rsyslog configuration would not need network online connectivity for simple local syslog logging. And this would delay syslog logging until the network is up - or even break it if network never comes up. 

I think the solution to this problem is more difficult.

**Comment by schlitzered** [comment]
please be aware that the standard centos, as well as amazon linux 2, contains these dependencies.

here is what the stock systemd unit file on centos 7 looks like:

``Description=System Logging Service
;Requires=syslog.socket
Wants=network.target network-online.target
After=network.target network-online.target
Documentation=man:rsyslogd(8)
Documentation=http://www.rsyslog.com/doc/

[Service]
Type=notify
EnvironmentFile=-/etc/sysconfig/rsyslog
ExecStart=/usr/sbin/rsyslogd -n $SYSLOGD_OPTIONS
Restart=on-failure
UMask=0066
StandardOutput=null
Restart=on-failure

[Install]
WantedBy=multi-user.target
;Alias=syslog.service
```

i found this issue, because cloud-init usually changes the hostname of a system, but rsyslog is not picking this up, because it is startet before cloud init finished.

i would also doubt that one will loose syslog messages with this. because applications using the syslog socket will simply block, because nothing is picking up messages from the log socket.

IMHO: i see no harm adding these dependencies

**Comment by ebsf1** [comment]
> this would delay syslog logging until the network is up - or even break it if network never comes up.

This occurred to me, too, but imrelp doesn't bind to its port (see https://github.com/rsyslog/rsyslog/issues/4293) and certain filters in imuxsock rulesets fail (see https://github.com/rsyslog/rsyslog/issues/4299) without it.

The alternative, at least for local logging, may be imjournal, which is fraught.

> I think the solution to this problem is more difficult.

Much must depend on syslog.socket.

**Comment by ebsf1** [comment]
> rsyslog/rsyslog#1656 may be potentially relevant here.
> 
> See also:
> 
> https://github.com/deoren/rsyslog-examples/blob/master/etc/systemd/system/rsyslog.service.d/10-wait-on-network.conf

@atc0005 agreed--at first blush, these appear quite similar.

**Comment by alorbach** [comment]
I agree that In shortterm, this solution is ok. 

In longerm I think rsyslog / librelp may should use "IP_FREEBIND" socket option to bind on network address that are nonlocal or do not (yet) exist. IP_FREEBIND is used in net.c / imudp already. 

@rgerhards what do you think?

**Comment by rgerhards** [comment]
> @rgerhards what do you think?

I have already begun to roll part of this to the OBS packages. My main concern is that we need to stick with the distro defaults. After all, our packages should be drop-in replacements, so we should change, at the OS config level, as little as possible.

I am also changing the default, for reasons given here: https://github.com/rsyslog/rsyslog/pull/4317
(sorry for forgetting the xref).

Anyways, I think if you can make this work with IP_FREEBIND without the network active, I think that would be a plus. At least I wouldn't argue against merging such a patch.

**Comment by rgerhards** [comment]
> filters in imuxsock rulesets fail (see [rsyslog/rsyslog#4299](https://github.com/rsyslog/rsyslog/issues/4299)) without it.

I think this is an unrelated observation

**Comment by ebsf1** [comment]
>     filters in imuxsock rulesets fail (see rsyslog/rsyslog#4299) without it.
> 
> I think this is an unrelated observation

@rgerhards this doubtless is true from an architectural standpoint.  I can easily imagine that changing the librelp socket option would have no effect on whether or how imuxsock rules and filters operate.

Still, it remains that these filters require this systemd dependency (and other filters fail altogether, regardless).  Another mechanism likely is at work (or not) but shares this characteristic.

**Comment by rgerhards** [comment]
> Still, it remains that these filters require this systemd dependency (and other filters fail altogether, regardless). Another mechanism likely is at work (or not) but shares this characteristic.

you mean your imuxsock issue with filtering went away when you applied the change to start rsyslog after the network?

**Comment by davidelang** [comment]
I think that would be a much better option.

**Comment by ebsf1** [comment]
> 
> 
>     Still, it remains that these filters require this systemd dependency (and other filters fail altogether, regardless). Another mechanism likely is at work (or not) but shares this characteristic.
> 
> you mean your imuxsock issue with filtering went away when you applied the change to start rsyslog after the network?

@rgerhards , yes, exactly, but only partially.  Adding the dependency enables some rules but not others (and, (i) all of the rules function without the dependency when not in an imuxsock ruleset (ii) syntactically identical rules function without the dependency in an imudp ruleset).

See https://github.com/rsyslog/rsyslog/issues/4299, which I opened 15 days ago and revised 11 days ago.  

Also see my comment from four days ago in https://github.com/rsyslog/rsyslog/issues/4293, where I detail which imuxsock rules succeed after adding the dependency, and which ones continue to fail.

Specifically, the rules that are enabled have one of the two following filters:
`if $programname == "rsyslogd" then {`
or
`:programname,isequal,"rsyslogd" `

The rules that continue to fail have one of the two following filters:
`if $msg contains "iptb" then {`
or 
`:msg,contains,"iptb"`

I don't know whether what matters is the property name or the comparison operator, because both are different and I only took my A/B testing so far.

I posted the rsyslog.conf file and the debug file already, between the two issues.

**Comment by ebsf1** [comment]
@rgerhards , I also observed that the rule with the selector/property-based filter (_i.e._, :msg,contiains,"iptb") implements the imuxsock module parameters for template but not the imuxsock module parameters for file or directory, regardless of whether the rule is in the imuxsock ruleset or free-standing, and regardless of the dependency.

**Comment by rgerhards** [comment]
> @rgerhards , yes, exactly, but only partially. Adding the dependency enables some rules but not others

This *is* unrelated, there is no logical reason it has to do with time of startup

**Comment by schlitzered** [comment]
is there an ETA when new packages with fixed unit files will be released?

**Comment by rgerhards** [comment]
Goal is with the next scheduled stable at last, which will most probably be next Tuesday. Daily stable may get them earlier.

**Comment by rgerhards** [comment]
Just FYI: This is already fixed in the OBS repo.

**Comment by rgerhards** [comment]
fixed in 8.2006.0, but pls see https://github.com/rsyslog/rsyslog/issues/4333.
